### PR TITLE
Feature/write test for ensembl uploadfile

### PR DESCRIPTION
**Pull request by mathieulemieux**
This script test the ensembl loader uploadFile function.

A .tsv export file from Biomart for gene ENSG00000139618 is used as input file for the loader.
Requests to the GraphKB's ApiConnection request method has been recorded and saved as .json for mocking.
Requests to externals APIs via node-fetch has been recorded and added to external_api_responses.json for mocking.

All expected newly created database objects are described in a constant.
The suite tests expected requests against calls on spied-on ApiConnection request method.

### CODE REVIEW

File: test/ensembl/ensembl.uploadFile.test.js [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    constructor() {
+        super();
+        // Data in mockfile are ordered to match actual request order
+        this.mockFile = path.join(__dirname, '../data/ensembl_uploadFile_requests.json');
+        this.mockDataset = JSON.parse(fs.readFileSync(this.mockFile));
+    }
+
+    request() {
+        const { result } = this.mockDataset[0];
+        this.mockDataset.shift();
```
Comment (elewis2) [inline test/ensembl/ensembl.uploadFile.test.js:27]: i believe shift() returns array[0] so you can just do const { result } = this.mockDataset.shift()

**Review by elewis2 — DISMISSED** [review]

File: test/ensembl/ensembl.uploadFile.test.js (hunk 1 above)
Comment (mathieulemieux) [reply test/ensembl/ensembl.uploadFile.test.js:27]: yes, totally. I'll change that

**Comment by mathieulemieux** [comment]
Had  to run "npx install-peerdeps --dev eslint-config-airbnb' as per documentation: https://www.npmjs.com/package/eslint-config-airbnb in order to have all GH checks passed.

**Review by elewis2 — APPROVED** [review]

File: test/ensembl/ensembl.uploadFile.test.js [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+class MockApiConnection extends ApiConnection {
+    constructor() {
+        super();
+        // Data in mockfile are ordered to match actual request order
+        this.mockFile = path.join(__dirname, '../data/ensembl_uploadFile_requests.json');
+        this.mockDataset = JSON.parse(fs.readFileSync(this.mockFile));
+    }
+
+    request() {
+        const { result } = this.mockDataset.shift();
```
Comment (kttkjl) [inline test/ensembl/ensembl.uploadFile.test.js:26]: curious, what does `.shift()` here do?

File: test/ensembl/ensembl.uploadFile.test.js [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+    ].map(el => Object.assign(el, {
+        // Custom method - Returns a string representation of the feature
+        toString: () => ('name' in el.feature
+            ? `${el.feature.biotype} Feature ${el.feature.name} from ${el.feature.source}`
+            : `${el.feature.biotype} Feature ${el.feature.sourceId} from ${el.feature.source}`),
+    })).map(el => Object.assign(el, {
+        // Custom method - Returns the source's RID from mockDataset
+        getSourceRid: () => ((conn.request).mock.results.filter(
+            (a) => a.value.result.name === el.feature.source,
+        )[0].value.result['@rid']),
```
Comment (kttkjl) [inline test/ensembl/ensembl.uploadFile.test.js:216]: so far the test dataset only has 1 result, would this be a problem when it is not of length 1?

**Review by kttkjl — COMMENTED** [review]
Not an expert on this, mostly asking for clarification

File: test/ensembl/ensembl.uploadFile.test.js [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+    request() {
+        const { result } = this.mockDataset.shift();
+        return result;
+    }
+}
+
+describe('uploadFile in Ensembl loader', () => {
+    const conn = new MockApiConnection(),
+        filename = path.join(__dirname, '../data/ensembl_biomart_export_ENSG00000139618.tsv');
```
[the remark this answers is not in the document]
Comment (mathieulemieux) [reply test/ensembl/ensembl.uploadFile.test.js:33]: The linter config we are using enforces this on save unfortunately.

File: test/ensembl/ensembl.uploadFile.test.js (hunk 2 above)
Comment (mathieulemieux) [reply test/ensembl/ensembl.uploadFile.test.js:26]: I will add a comment in the file to be more explicit about it; thank you @kttkjl for pointing it out.

File: test/ensembl/ensembl.uploadFile.test.js (hunk 3 above)
Comment (mathieulemieux) [reply test/ensembl/ensembl.uploadFile.test.js:216]: When testing the loader for adding the right Feature objects to the database, we use the getSourceRid() method to make sure the Feature has been given the right Source RID as Source property value: https://github.com/bcgsc/pori_graphkb_loader/blob/5a6a7bba4e8d37071f0788c24d496826f37477a9/test/ensembl/ensembl.uploadFile.test.js#L236

This is why every object of the 'expected' list has been given the same method allowing them to convert their source from name to RID. The correct RID is obtained from returned values from the Jest mock object. If more than one result returned the same name, then it should also return the same RID, so we can juste grab the first returned result : https://github.com/bcgsc/pori_graphkb_loader/blob/5a6a7bba4e8d37071f0788c24d496826f37477a9/test/ensembl/ensembl.uploadFile.test.js#L216

Same for the edges here:
https://github.com/bcgsc/pori_graphkb_loader/blob/5a6a7bba4e8d37071f0788c24d496826f37477a9/test/ensembl/ensembl.uploadFile.test.js#L249

If we want to make sure that the name of a source is not also the name of an other object returned by the mock dataset, then we could also add a @class filter. I'll take care of adding that; thanks @kttkjl for the comment.

File: test/ensembl/ensembl.uploadFile.test.js [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+                biotype: 'transcript', source: 'refseq', sourceId: 'NM_000059', sourceIdVersion: null,
+            },
+
+        },
+    ].map(el => Object.assign(el, {
+        // Custom method - Returns a string representation of the feature
+        toString: () => ('name' in el.feature
+            ? `${el.feature.biotype} Feature ${el.feature.name} from ${el.feature.source}`
+            : `${el.feature.biotype} Feature ${el.feature.sourceId} from ${el.feature.source}`),
+    })).map(el => Object.assign(el, {
```
[the remark this answers is not in the document]
Comment (mathieulemieux) [reply test/ensembl/ensembl.uploadFile.test.js:212]: Totally! I'll fix that.

File: test/ensembl/ensembl.uploadFile.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (kttkjl) [reply test/ensembl/ensembl.uploadFile.test.js:34]: oh that's unfortunate ... weird I thought normal linters would be against this sort of assignment O_O

File: test/ensembl/ensembl.uploadFile.test.js (hunk 3 above)
Comment (kttkjl) [reply test/ensembl/ensembl.uploadFile.test.js]: Ohh I see, thanks for clearing that up!

File: test/ensembl/ensembl.uploadFile.test.js (hunk 2 above)
Comment (kttkjl) [reply test/ensembl/ensembl.uploadFile.test.js:27]: I see, it does seem a bit confusing at first, but I understand the sentiment

**Review by kttkjl — APPROVED** [review]

**Pull request merged**